# data_p_sub.csv

In [64]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [37]:
df = pd.read_csv("data-p3/dat_P_sub_c.csv")

In [38]:
with pd.option_context('display.max_columns', None):
    print(df.head())

   Pass_ID  valid_P_ID Airfield               S2  Wait_Time  C_Start  C0  \
0  5348206           1      AUC  2028-09-01 6:37        NaN        1   1   
1  5348209           1      AUC  2028-09-01 6:38        NaN        1   1   
2  5348216           1      AUC  2028-09-01 6:39        NaN        1   1   
3  5348224           1      AUC  2028-09-01 6:39        NaN        1   1   
4  5348225           1      AUC  2028-09-01 6:40        NaN        1   1   

   C_avg    Sch_Departure    Act_Departure BFO_Dest_City  \
0    1.0  2028-09-01 8:06  2028-09-01 8:06        BORQUE   
1    1.0  2028-09-01 8:06  2028-09-01 8:06        BORQUE   
2    1.0  2028-09-01 8:06  2028-09-01 8:06        BORQUE   
3    1.0  2028-09-01 8:06  2028-09-01 8:06        BORQUE   
4    1.0  2028-09-01 8:06  2028-09-01 8:06        BORQUE   

  BFO_Destination_Country_Code    order Departure_Date  Time_of_Day  \
0                          BOR  5374650     2028-09-01  2 - MORNING   
1                          BOR  5374657 

In [39]:
df.dtypes

Pass_ID                           int64
valid_P_ID                        int64
Airfield                         object
S2                               object
Wait_Time                       float64
C_Start                           int64
C0                                int64
C_avg                           float64
Sch_Departure                    object
Act_Departure                    object
BFO_Dest_City                    object
BFO_Destination_Country_Code     object
order                             int64
Departure_Date                   object
Time_of_Day                      object
Period_of_Week                   object
Day_of_Week                      object
Month                            object
Season                           object
Year                              int64
WT_flag                           int64
S2_Sch_Flag                       int64
S2_Act_Flag                       int64
Sch_Act_Flag                      int64
Flight_ID                         int64


In [40]:
# dates = ['S2','Sch_Departure', 'Act_Departure','Departure_Date','Year']
# df[dates] = df[dates].apply(pd.to_datetime)

df['S2'] = pd.to_datetime(df['S2']).dt.to_period('min')
df['Sch_Departure'] = pd.to_datetime(df['Sch_Departure']).dt.to_period('min')
df['Act_Departure'] = pd.to_datetime(df['Act_Departure']).dt.to_period('min')

# Convert to day periods
df['Departure_Date'] = pd.to_datetime(df['Departure_Date']).dt.to_period('D')

# Convert to year periods
df['Year'] = pd.to_datetime(df['Year'], format='%Y').dt.to_period('Y')

In [41]:
df.head()

Pass_ID  valid_P_ID Airfield                S2  Wait_Time  C_Start  C0  \
0  5348206           1      AUC  2028-09-01 06:37        NaN        1   1   
1  5348209           1      AUC  2028-09-01 06:38        NaN        1   1   
2  5348216           1      AUC  2028-09-01 06:39        NaN        1   1   
3  5348224           1      AUC  2028-09-01 06:39        NaN        1   1   
4  5348225           1      AUC  2028-09-01 06:40        NaN        1   1   

   C_avg     Sch_Departure     Act_Departure  ... Day_of_Week     Month  \
0    1.0  2028-09-01 08:06  2028-09-01 08:06  ...     5 - FRI  09 - SEP   
1    1.0  2028-09-01 08:06  2028-09-01 08:06  ...     5 - FRI  09 - SEP   
2    1.0  2028-09-01 08:06  2028-09-01 08:06  ...     5 - FRI  09 - SEP   
3    1.0  2028-09-01 08:06  2028-09-01 08:06  ...     5 - FRI  09 - SEP   
4    1.0  2028-09-01 08:06  2028-09-01 08:06  ...     5 - FRI  09 - SEP   

       Season  Year WT_flag S2_Sch_Flag S2_Act_Flag Sch_Act_Flag Flight_ID  \
0  3 - SUMMER  2028       1           0           0            0     18095   
1  3 - SUMMER  2028       1           0           0            0     18095   
2  3 - SUMMER  2028       1           0           0            0     18095   
3  3 - SUMMER  2028       1           0           0            0     18095   
4  3 - SUMMER  2028       1           0           0            0     18095   

  Delay_in_Seconds  
0                0  
1                0  
2                0  
3                0  
4                0  

[5 rows x 26 columns]

In [42]:
df.dtypes

Pass_ID                                 int64
valid_P_ID                              int64
Airfield                               object
S2                                period[min]
Wait_Time                             float64
C_Start                                 int64
C0                                      int64
C_avg                                 float64
Sch_Departure                     period[min]
Act_Departure                     period[min]
BFO_Dest_City                          object
BFO_Destination_Country_Code           object
order                                   int64
Departure_Date                      period[D]
Time_of_Day                            object
Period_of_Week                         object
Day_of_Week                            object
Month                                  object
Season                                 object
Year                            period[Y-DEC]
WT_flag                                 int64
S2_Sch_Flag                       

In [43]:
df['BFO_Destination_Country_Code'].value_counts()

BFO_Destination_Country_Code
BOR    82740
VES    22955
SCO     5673
NEN     2751
  .        8
WIC        5
Name: count, dtype: int64

In [44]:
with pd.option_context('display.max_columns', None):
    print(df[df["BFO_Destination_Country_Code"] == "  ."])

        Pass_ID  valid_P_ID Airfield                S2  Wait_Time  C_Start  \
1472    5368296           1      AUC  2028-09-03 11:26        3.0        1   
11379   5480057           1      AUC  2028-09-15 07:45        3.0        1   
27518   5648611           1      AUC  2028-10-03 08:24        2.0        2   
52319   5888442           1      AUC  2028-10-28 06:43        6.0        2   
96020   6290973           1      AUC  2028-12-15 20:10        NaN        1   
96654   6296559           1      AUC  2028-12-16 14:45        3.0        2   
109222  6402157           1      AUC  2028-12-28 07:18       14.0        2   
113148  6427957           1      AUC  2028-12-30 18:49       21.0        3   

        C0  C_avg     Sch_Departure     Act_Departure BFO_Dest_City  \
1472     1    1.0  2028-09-03 13:07  2028-09-03 13:07             .   
11379    1    1.0  2028-09-15 09:31  2028-09-15 09:31             .   
27518    2    2.0  2028-10-03 09:36  2028-10-03 13:06             .   
52319    2   

In [45]:
df['BFO_Dest_City'] = df['BFO_Dest_City'].replace("  .", np.nan)
df['BFO_Destination_Country_Code'] = df['BFO_Destination_Country_Code'].replace("  .", np.nan)

In [46]:
df

Pass_ID  valid_P_ID Airfield                S2  Wait_Time  C_Start  \
0       5348206           1      AUC  2028-09-01 06:37        NaN        1   
1       5348209           1      AUC  2028-09-01 06:38        NaN        1   
2       5348216           1      AUC  2028-09-01 06:39        NaN        1   
3       5348224           1      AUC  2028-09-01 06:39        NaN        1   
4       5348225           1      AUC  2028-09-01 06:40        NaN        1   
...         ...         ...      ...               ...        ...      ...   
114127  6438181           1      AUC  2028-12-31 18:55        NaN        1   
114128  6438424           1      AUC  2028-12-31 19:18        NaN        1   
114129  6438426           1      AUC  2028-12-31 19:18        NaN        1   
114130  6438427           1      AUC  2028-12-31 19:19        NaN        1   
114131  6438439           1      AUC  2028-12-31 19:21        NaN        1   

        C0  C_avg     Sch_Departure     Act_Departure  ... Day_of_Week  \
0        1    1.0  2028-09-01 08:06  2028-09-01 08:06  ...     5 - FRI   
1        1    1.0  2028-09-01 08:06  2028-09-01 08:06  ...     5 - FRI   
2        1    1.0  2028-09-01 08:06  2028-09-01 08:06  ...     5 - FRI   
3        1    1.0  2028-09-01 08:06  2028-09-01 08:06  ...     5 - FRI   
4        1    1.0  2028-09-01 08:06  2028-09-01 08:06  ...     5 - FRI   
...     ..    ...               ...               ...  ...         ...   
114127   1    1.0  2028-12-31 20:06  2028-12-31 20:06  ...     7 - SUN   
114128   1    1.0  2028-12-31 20:06  2028-12-31 20:06  ...     7 - SUN   
114129   1    1.0  2028-12-31 20:06  2028-12-31 20:06  ...     7 - SUN   
114130   1    1.0  2028-12-31 20:06  2028-12-31 20:06  ...     7 - SUN   
114131   1    1.0  2028-12-31 20:06  2028-12-31 20:06  ...     7 - SUN   

           Month      Season  Year WT_flag S2_Sch_Flag S2_Act_Flag  \
0       09 - SEP  3 - SUMMER  2028       1           0           0   
1       09 - SEP  3 - SUMMER  2028       1           0           0   
2       09 - SEP  3 - SUMMER  2028       1           0           0   
3       09 - SEP  3 - SUMMER  2028       1           0           0   
4       09 - SEP  3 - SUMMER  2028       1           0           0   
...          ...         ...   ...     ...         ...         ...   
114127  12 - DEC  4 - AUTUMN  2028       1           0           0   
114128  12 - DEC  4 - AUTUMN  2028       1           0           0   
114129  12 - DEC  4 - AUTUMN  2028       1           0           0   
114130  12 - DEC  4 - AUTUMN  2028       1           0           0   
114131  12 - DEC  4 - AUTUMN  2028       1           0           0   

       Sch_Act_Flag Flight_ID Delay_in_Seconds  
0                 0     18095                0  
1                 0     18095                0  
2                 0     18095                0  
3                 0     18095                0  
4                 0     18095                0  
...             ...       ...              ...  
114127            0     21678                0  
114128            0     21678                0  
114129            0     21678                0  
114130            0     21678                0  
114131            0     21678                0  

[114132 rows x 26 columns]

In [52]:
low_cardinality_int_cols = [col for col in df.select_dtypes(include=['int']).columns if df[col].nunique() < 10]
object_cols = df.select_dtypes(include=['object']).columns.tolist()
high_cardinality_int_cols = [col for col in df.select_dtypes(include=['int']).columns if df[col].nunique() >= 10]
continuous_cols = df.select_dtypes(include=['float']).columns.tolist()
period_cols = [col for col in df.columns if pd.api.types.is_period_dtype(df[col])]

dtypes = {
    "Column": [],
    "Type": [],
    "Domain": [],
    "Missing": []
}

for col in low_cardinality_int_cols + object_cols:
    unique_vals = df[col].dropna().unique()
    dtypes["Column"].append(col)
    dtypes["Type"].append("Low Cardinality Integer" if col in low_cardinality_int_cols else "Object")
    dtypes["Domain"].append(set(unique_vals))
    dtypes["Missing"].append(df[col].isna().sum())

for col in high_cardinality_int_cols + continuous_cols:
    min_val = df[col].min()
    max_val = df[col].max()
    dtypes["Column"].append(col)
    dtypes["Type"].append("High Cardinality Integer" if col in high_cardinality_int_cols else "Continuous")
    dtypes["Domain"].append([min_val, max_val])
    dtypes["Missing"].append(df[col].isna().sum())

for col in period_cols:
    min_val = df[col].min()
    max_val = df[col].max()
    dtypes["Column"].append(col)
    dtypes["Type"].append("Period")
    dtypes["Domain"].append(set([min_val]) if min_val == max_val else [min_val, max_val])
    dtypes["Missing"].append(df[col].isna().sum())

df_dtypes = pd.DataFrame(dtypes)
df_dtypes


C:\Users\mairi\AppData\Local\Temp\ipykernel_22572\4005902089.py:5: DeprecationWarning: is_period_dtype is deprecated and will be removed in a future version. Use `isinstance(dtype, pd.PeriodDtype)` instead
  period_cols = [col for col in df.columns if pd.api.types.is_period_dtype(df[col])]


Column                      Type  \
0                     valid_P_ID   Low Cardinality Integer   
1                        C_Start   Low Cardinality Integer   
2                             C0   Low Cardinality Integer   
3                        WT_flag   Low Cardinality Integer   
4                    S2_Sch_Flag   Low Cardinality Integer   
5                    S2_Act_Flag   Low Cardinality Integer   
6                   Sch_Act_Flag   Low Cardinality Integer   
7                       Airfield                    Object   
8                  BFO_Dest_City                    Object   
9   BFO_Destination_Country_Code                    Object   
10                   Time_of_Day                    Object   
11                Period_of_Week                    Object   
12                   Day_of_Week                    Object   
13                         Month                    Object   
14                        Season                    Object   
15                       Pass_ID  High Cardinality Integer   
16                         order  High Cardinality Integer   
17                     Flight_ID  High Cardinality Integer   
18              Delay_in_Seconds  High Cardinality Integer   
19                     Wait_Time                Continuous   
20                         C_avg                Continuous   
21                            S2                    Period   
22                 Sch_Departure                    Period   
23                 Act_Departure                    Period   
24                Departure_Date                    Period   
25                          Year                    Period   

                                               Domain  Missing  
0                                                 {1}        0  
1                                           {1, 2, 3}        0  
2                                           {1, 2, 3}        0  
3                                              {0, 1}        0  
4                                              {0, 1}        0  
5                                              {0, 1}        0  
6                                              {0, 1}        0  
7                                               {AUC}        0  
8   {VES009, SCO002, NEN003, VES034, VES056, VES07...        0  
9                           {VES, BOR, NEN, WIC, SCO}        8  
10  {1 - NIGHT, 4 - EVENING, 3 - AFTERNOON, 2 - MO...        0  
11                         {2 - WEEKEND, 1 - WEEKDAY}        0  
12  {7 - SUN, 3 - WED, 2 - TUE, 1 - MON, 6 - SAT, ...        0  
13           {12 - DEC, 09 - SEP, 10 - OCT, 11 - NOV}        0  
14                           {3 - SUMMER, 4 - AUTUMN}        0  
15                                 [5348206, 6438752]        0  
16                                 [5374650, 6459925]        0  
17                                     [18095, 21678]        0  
18                                    [-1440, 100200]        0  
19                                        [1.0, 75.0]    16871  
20                                         [0.5, 3.0]        0  
21               [2028-09-01 06:37, 2028-12-31 20:26]        0  
22               [2028-09-01 08:06, 2028-12-31 20:06]        0  
23               [2028-09-01 08:06, 2028-12-31 21:40]        0  
24                           [2028-09-01, 2028-12-31]        0  
25                                             {2028}        0

In [57]:
low_cardinality_int_cols = [col for col in df.select_dtypes(include=['int']).columns if df[col].nunique() < 3]
object_cols = df.select_dtypes(include=['object']).columns.tolist()
# for col in object_cols:
#     df[col] = df[col].astype('string')
high_cardinality_int_cols = [col for col in df.select_dtypes(include=['int']).columns if df[col].nunique() >= 3]
continuous_cols = df.select_dtypes(include=['float']).columns.tolist()
period_cols = [col for col in df.columns if pd.api.types.is_period_dtype(df[col])]

dtypes = {
    "Column": [],
    "Type": [],
    "Domain": [],
}

for col in low_cardinality_int_cols + object_cols:
    unique_vals = df[col].dropna().unique()
    dtypes["Column"].append(col)
    dtypes["Type"].append('String' if df[col].dtype == 'object' else 'Boolean')
    dtypes["Domain"].append(set(unique_vals))

for col in high_cardinality_int_cols + continuous_cols:
    min_val = df[col].min()
    max_val = df[col].max()
    dtypes["Column"].append(col)
    dtypes["Type"].append(f'{df[col].dtype}'.replace("int64","Integer").replace("float64","Float"))
    dtypes["Domain"].append([min_val, max_val])
   

for col in period_cols:
    min_val = df[col].min()
    max_val = df[col].max()
    dtypes["Column"].append(col)
    dtypes["Type"].append("DateTime")
    dtypes["Domain"].append(set([min_val]) if min_val == max_val else [min_val, max_val])

df_dtypes = pd.DataFrame(dtypes)
df_dtypes


C:\Users\mairi\AppData\Local\Temp\ipykernel_22572\830724953.py:7: DeprecationWarning: is_period_dtype is deprecated and will be removed in a future version. Use `isinstance(dtype, pd.PeriodDtype)` instead
  period_cols = [col for col in df.columns if pd.api.types.is_period_dtype(df[col])]


Column      Type  \
0                     valid_P_ID   Boolean   
1                        WT_flag   Boolean   
2                    S2_Sch_Flag   Boolean   
3                    S2_Act_Flag   Boolean   
4                   Sch_Act_Flag   Boolean   
5                       Airfield    String   
6                  BFO_Dest_City    String   
7   BFO_Destination_Country_Code    String   
8                    Time_of_Day    String   
9                 Period_of_Week    String   
10                   Day_of_Week    String   
11                         Month    String   
12                        Season    String   
13                       Pass_ID   Integer   
14                       C_Start   Integer   
15                            C0   Integer   
16                         order   Integer   
17                     Flight_ID   Integer   
18              Delay_in_Seconds   Integer   
19                     Wait_Time     Float   
20                         C_avg     Float   
21                            S2  DateTime   
22                 Sch_Departure  DateTime   
23                 Act_Departure  DateTime   
24                Departure_Date  DateTime   
25                          Year  DateTime   

                                               Domain  
0                                                 {1}  
1                                              {0, 1}  
2                                              {0, 1}  
3                                              {0, 1}  
4                                              {0, 1}  
5                                               {AUC}  
6   {VES009, SCO002, NEN003, VES034, VES056, VES07...  
7                           {VES, BOR, NEN, WIC, SCO}  
8   {1 - NIGHT, 4 - EVENING, 3 - AFTERNOON, 2 - MO...  
9                          {2 - WEEKEND, 1 - WEEKDAY}  
10  {7 - SUN, 3 - WED, 2 - TUE, 1 - MON, 6 - SAT, ...  
11           {12 - DEC, 09 - SEP, 10 - OCT, 11 - NOV}  
12                           {3 - SUMMER, 4 - AUTUMN}  
13                                 [5348206, 6438752]  
14                                             [1, 3]  
15                                             [1, 3]  
16                                 [5374650, 6459925]  
17                                     [18095, 21678]  
18                                    [-1440, 100200]  
19                                        [1.0, 75.0]  
20                                         [0.5, 3.0]  
21               [2028-09-01 06:37, 2028-12-31 20:26]  
22               [2028-09-01 08:06, 2028-12-31 20:06]  
23               [2028-09-01 08:06, 2028-12-31 21:40]  
24                           [2028-09-01, 2028-12-31]  
25                                             {2028}

In [87]:
# Convert Period columns to datetime using .to_timestamp() instead of pd.to_datetime
s2_datetime = df['S2'].apply(lambda x: x.to_timestamp() if pd.notnull(x) else pd.NaT)
sch_departure_datetime = df['Sch_Departure'].apply(lambda x: x.to_timestamp() if pd.notnull(x) else pd.NaT)
act_departure_datetime = df['Act_Departure'].apply(lambda x: x.to_timestamp() if pd.notnull(x) else pd.NaT)

# Calculate time differences in minutes, handling NaNs by setting invalid differences to NaN
sch_act_diff = (act_departure_datetime - sch_departure_datetime).dt.total_seconds() / 60
s2_sch_diff = (sch_departure_datetime - s2_datetime).dt.total_seconds() / 60
s2_act_diff = (act_departure_datetime - s2_datetime).dt.total_seconds() / 60

# Create a temporary DataFrame with these calculated differences for analysis
temp_df = pd.DataFrame({
    'Sch_Act_Flag': df['Sch_Act_Flag'],
    'S2_Sch_Flag': df['S2_Sch_Flag'],
    'S2_Act_Flag': df['S2_Act_Flag'],
    'Sch_Act_Diff': sch_act_diff,
    'S2_Sch_Diff': s2_sch_diff,
    'S2_Act_Diff': s2_act_diff
})

# Filter out rows with NaN values in the difference columns before performing min and max calculations
temp_df = temp_df.dropna(subset=['Sch_Act_Diff', 'S2_Sch_Diff', 'S2_Act_Diff'])

# Calculate min and max of time differences by flag values
sch_act_diff_by_flag_min = temp_df.groupby('Sch_Act_Flag')['Sch_Act_Diff'].min()
sch_act_diff_by_flag_max = temp_df.groupby('Sch_Act_Flag')['Sch_Act_Diff'].max()

s2_sch_diff_by_flag_min = temp_df.groupby('S2_Sch_Flag')['S2_Sch_Diff'].min()
s2_sch_diff_by_flag_max = temp_df.groupby('S2_Sch_Flag')['S2_Sch_Diff'].max()

s2_act_diff_by_flag_min = temp_df.groupby('S2_Act_Flag')['S2_Act_Diff'].min()
s2_act_diff_by_flag_max = temp_df.groupby('S2_Act_Flag')['S2_Act_Diff'].max()

# Display results
print("Minimum actual departure - scheduled departure:\n", sch_act_diff_by_flag_min)
print("Maximum actual departure - scheduled departure:\n", sch_act_diff_by_flag_max)
print("\nMinimum scheduled departure - arrival at S2:\n", s2_sch_diff_by_flag_min)
print("Maximum scheduled departure - arrival at S2:\n", s2_sch_diff_by_flag_max)
print("\nMinimum actual departure - arrival at S2:\n", s2_act_diff_by_flag_min)
print("Maximum actual departure - arrival at S2:\n", s2_act_diff_by_flag_max)


Minimum actual departure - scheduled departure:
 Sch_Act_Flag
0     0.0
1   -24.0
Name: Sch_Act_Diff, dtype: float64
Maximum actual departure - scheduled departure:
 Sch_Act_Flag
0    1670.0
1      -1.0
Name: Sch_Act_Diff, dtype: float64

Minimum scheduled departure - arrival at S2:
 S2_Sch_Flag
0        0.0
1   -40303.0
Name: S2_Sch_Diff, dtype: float64
Maximum scheduled departure - arrival at S2:
 S2_Sch_Flag
0    1408.0
1      -1.0
Name: S2_Sch_Diff, dtype: float64

Minimum actual departure - arrival at S2:
 S2_Act_Flag
0        0.0
1   -40303.0
Name: S2_Act_Diff, dtype: float64
Maximum actual departure - arrival at S2:
 S2_Act_Flag
0    1756.0
1      -1.0
Name: S2_Act_Diff, dtype: float64


| Column                          | Type       | Description                                           |
|---------------------------------|------------|--------------------------------------------------|
| valid_P_ID                      | Boolean    | Boolean representing whether the passenger's ID is valid or not (0 for no, 1 for yes). All passengers in this dataset have valid IDs.|
| WT_flag                         | Boolean    | Boolean representing whether Wait_Time is NA (1 for yes, 0 for no).|                                          |
| S2_Sch_Flag                     | Boolean    | Boolean representing whether the passenger arrived at S2 after their plane was scheduled to leave (1 for yes, two for no). |
| S2_Act_Flag                     | Boolean    | Boolean representing whether the passenger arrived at S2 after their plane left (1 for yes, two for no).                                          |
| Sch_Act_Flag                    | Boolean    | Boolean representing whether the plane left early (1 for yes, two for no).                                           |
| Airfield                        | String     | The airfield from which the flight departed; all entries in this dataset are AUC.                                            |
| BFO_Dest_City                   | String     | Destination city of passenger's flight.|
| BFO_Destination_Country_Code    | String     | Destination country of passenger's flight. Values in dataset are VES, BOR, NEN, WIC, SCO.|
| Time_of_Day                     | String     | Flight's time of day (morning, afternoon, evening, or night).|
| Period_of_Week                  | String     | Whether the flight was on a weekday or weekend.                     |
| Day_of_Week                     | String     | The day of the week that the flight took place.|
| Month                           | String     | The month, ranging from September to December, that the flight took place.        |
| Season                          | String     | The season (summer or autumn) that the flight took place                       |
| Pass_ID                         | Integer    | The passenger's pass ID number, ranging from 5348206 to 6438752.                           |
| C_Start                         | Integer    | The number of servers at S1 when the passenger arrived.                                          |
| C0                              | Integer    | The number servers at S2 when the passenger reached the front of the line.                                           |
| order                           | Integer    | The passenger's order in line, ranging from 5374650 to 6459925.                            |
| Flight_ID                       | Integer    | The flight ID number, ranging from 18095 to 21678.                                  |
| Delay_in_Seconds                | Integer    | The difference between the scheduled departure time and the actual departure time in seconds, ranging from -1440 to 100200. Negative values indicate early departure. |
| Wait_Time                       | Float      | Wait time in minutres, ranging from 1.0 to 75.0                                      |
| C_avg                           | Float      | The average number of servers present at each gate while the passenger was in the queue, ranging from 0.5 to 3.  |
| S2                              | DateTime   | Arrival time at S2, between 2028-09-01 06:37 and 2028-12-31 20:26.|
| Sch_Departure                   | DateTime   | Scheduled departure time, between 2028-09-01 08:06 and 2028-12-31 20:06.           |
| Act_Departure                   | DateTime   | Actual departure time, between 2028-09-01 08:06 and 2028-12-31 21:40.  |
| Departure_Date                  | DateTime   | Day of flight departure, between 2028-09-01 and 2028-12-31.                        |
| Year                            | DateTime   | Year of flight departure; all entries are 2028.                             
